core/consensus/authority/impl/authority_manager_impl.cpp

```diff
@@ -174,25 +179,51 @@ namespace kagome::authority {
         }
         const auto &header = header_res.value();
 
-        // observe possible changes of authorities
-        for (auto &digest : header.digest) {
-          visit_in_place(
-              digest,
-              [&](const primitives::Consensus &consensus_message) {
-                collected.emplace(ConsensusMessages{
-                    primitives::BlockInfo(header.number, hash),
-                    consensus_message});
-                if (consensus_message.consensus_engine_id
-                    == primitives::kGrandpaEngineId) {
-                  found = true;
-                }
-              },
-              [](const auto &...) {});  // Other variants are ignored
+        if (header.number == 0) {
+          found_set_change = true;
+        } else {
+          // observe possible changes of authorities
+          for (auto &digest : header.digest) {
+            visit_in_place(
+                digest,
+                [&](const primitives::Consensus &consensus_message) {
+                  collected.emplace(ConsensusMessages{
+                      primitives::BlockInfo(header.number, hash),
+                      consensus_message});
+                  bool is_grandpa = consensus_message.consensus_engine_id
+                                    == primitives::kGrandpaEngineId;
+                  auto decoded_res = consensus_message.decode();
+                  if (decoded_res.has_error()) {
+                    log_->critical(
+                        "Error decoding consensus digest message: {}",
+                        decoded_res.error().message());
+                    return;
+                  }
+                  if (is_grandpa) {
+                    bool is_scheduled_change =
+                        decoded_res.value()
+                            .isGrandpaDigestOf<primitives::ScheduledChange>();
+                    bool is_forced_change =
+                        decoded_res.value()
+                            .isGrandpaDigestOf<primitives::ForcedChange>();
+                    found_set_change = is_forced_change or is_scheduled_change;
```

Comment:
```suggestion
                    found_set_change = 
                      found_set_change or is_forced_change or is_scheduled_change;
```
Original code might cancel first result on next iterations

Reply:
Ah, right, we don't break here.